src/hffs/fs.py

```diff
@@ -118,35 +89,76 @@ def _resolve_repo_id(self, path: str) -> Tuple[str, str, str]:
             repo_type = huggingface_hub.constants.REPO_TYPES_MAPPING[repo_type]
         else:
             repo_type = huggingface_hub.constants.REPO_TYPE_MODEL
+        revision = None
         if path.count("/") > 0:
-            repo_id_with_namespace = "/".join(path.split("/")[:2])
-            path_in_repo_with_namespace = "/".join(path.split("/")[2:])
-            repo_id_without_namespace = path.split("/")[0]
-            path_in_repo_without_namespace = "/".join(path.split("/")[1:])
-            if self._repo_exists(repo_id_with_namespace, repo_type=repo_type):
-                repo_id = repo_id_with_namespace
-                path_in_repo = path_in_repo_with_namespace
-            elif self._repo_exists(repo_id_without_namespace, repo_type=repo_type):
-                repo_id = repo_id_without_namespace
-                path_in_repo = path_in_repo_without_namespace
+            if "@" in path:
+                repo_id, revision = path.split("@", 1)
+                if "/" in revision:
+                    revision, path_in_repo = revision.split("/", 1)
+                else:
+                    path_in_repo = ""
+                revision = unquote(revision)
+                if not self._repo_exists(repo_id, repo_type=repo_type):
+                    raise FileNotFoundError(f"No such repository: '{repo_id}'")
+                return repo_type, repo_id, revision, path_in_repo
             else:
-                raise FileNotFoundError(f"No such repository: '{repo_id_with_namespace}'")
+                revision = None
+                repo_id_with_namespace = "/".join(path.split("/")[:2])
+                path_in_repo_with_namespace = "/".join(path.split("/")[2:])
+                repo_id_without_namespace = path.split("/")[0]
+                path_in_repo_without_namespace = "/".join(path.split("/")[1:])
+                if self._repo_exists(repo_id_with_namespace, repo_type=repo_type):
+                    repo_id = repo_id_with_namespace
+                    path_in_repo = path_in_repo_with_namespace
+                elif self._repo_exists(repo_id_without_namespace, repo_type=repo_type):
+                    repo_id = repo_id_without_namespace
+                    path_in_repo = path_in_repo_without_namespace
+                else:
+                    raise FileNotFoundError(f"No such repository: '{repo_id_with_namespace}'")
         else:
+            if "@" in path:
+                path, revision = path.split("@", 1)
+                revision = unquote(revision)
+            else:
+                revision = None
             if self._repo_exists(path, repo_type=repo_type):
                 repo_id = path
                 path_in_repo = ""
             else:
                 # can't list repositories at the namespace level
                 raise NotImplementedError("Acces to repositories lists is not implemented.")
 
-        return repo_type, repo_id, path_in_repo
+        return repo_type, repo_id, revision, path_in_repo
```

Comment:
(personal opinion) Once return type become too long like this one I tend to group it in a dataclass. Especially since all 4 values are really tied together. Maybe a `ResolvedPath` object?

Feel free to ignore if you don't like it. Pattern design is often subjective anyway.